cmd/snap-update-ns/snap-update-ns.c

```diff
@@ -19,7 +19,299 @@
+		debug("The current profile has been updated.");
+	}
+	if (num_skipped == 0 && num_changed == 0) {
+		debug("Mount namespace of snap %s is already up-to-date.",
+		      snap_name);
+	}
+	return 0;
+}
+
+static void sc_show_mountinfo(struct mountinfo_entry *mi_entry)
+{
+	debug("\t\tid:           %d", mountinfo_entry_mount_id(mi_entry));
+	debug("\t\tparent-id:    %d", mountinfo_entry_parent_id(mi_entry));
+	debug("\t\troot:         %s", mountinfo_entry_root(mi_entry));
+	debug("\t\tmount-dir:    %s", mountinfo_entry_mount_dir(mi_entry));
+	debug("\t\tmount-opts:   %s", mountinfo_entry_mount_opts(mi_entry));
+	debug("\t\toptional:     %s",
+	      mountinfo_entry_optional_fields(mi_entry));
+	debug("\t\tfs-type:      %s", mountinfo_entry_fs_type(mi_entry));
+	debug("\t\tmount-source: %s", mountinfo_entry_mount_source(mi_entry));
+	debug("\t\tsuper-opts:   %s", mountinfo_entry_super_opts(mi_entry));
+}
+
+static struct mountinfo_entry *sc_find_mountinfo_by_id(struct mountinfo *mi,
+						       int mount_id)
+{
+	for (struct mountinfo_entry * mi_entry =
+	     first_mountinfo_entry(mi); mi_entry != NULL;
+	     mi_entry = next_mountinfo_entry(mi_entry)) {
+		if (mountinfo_entry_mount_id(mi_entry) == mount_id) {
+			return mi_entry;
+		}
+	}
+	return NULL;
+}
+
+static bool sc_should_act_on_change(const struct sc_mount_change *change)
+{
+	// Load the table of mount points that affect the current process.  We're
+	// doing this each time we are asked to mount something as it is safer than
+	// trying to keep track of what the kernel may be doing.
+	struct mountinfo __attribute__ ((cleanup(cleanup_mountinfo))) * mi =
+	    NULL;
+	const char *mnt_dir;
+	mi = parse_mountinfo(NULL);
+
+	switch (change->action) {
+	case SC_ACTION_MOUNT:
+		// We cannot mount over existing mount points as that can confuse
+		// apparmor. As a safety measure we reject such mount requests.
+		for (struct mountinfo_entry * mi_entry =
+		     first_mountinfo_entry(mi); mi_entry != NULL;
+		     mi_entry = next_mountinfo_entry(mi_entry)) {
+			mnt_dir = mountinfo_entry_mount_dir(mi_entry);
+			// XXX: it would be perfect if this could detect that we don't have
+			// to do anything but it is not an error. Specifically for the case
+			// of bind mounts that are already satisfied.
+			if (strcmp(mnt_dir, change->entry->mnt_dir) == 0) {
+				debug("\tIgnoring request to mount over"
+				      " an existing mount-point: %s",
+				      change->entry->mnt_dir);
+				debug("\tIn the way:");
+				sc_show_mountinfo(mi_entry);
+				struct mountinfo_entry *parent_mi_entry =
+				    sc_find_mountinfo_by_id(mi,
+							    mountinfo_entry_parent_id
+							    (mi_entry));
+				while (parent_mi_entry != NULL) {
+					debug("\t(parent chain)...");
+					sc_show_mountinfo(parent_mi_entry);
+					parent_mi_entry =
+					    sc_find_mountinfo_by_id(mi,
+								    mountinfo_entry_parent_id
+								    (parent_mi_entry));
+				}
+				return false;
```

Comment:
I'll trim down the debug info if required (though we don't show it unless you ask for it). I'd like to return a simple and understandable error message when we reject the over-mount.